app/decorators/component.ts

```diff
@@ -0,0 +1,67 @@
+import Component from '@ember/component';
+import { assert } from '@ember/debug';
+
+/**
+ * Property decorator for required action arguments in components.
+ * Provides a default action method that raises a useful error.
+ *
+ * Usage:
+ *
+ * ```ts
+ *  import { requiredAction } from 'ember-osf-web/utils/required-action';
+ *
+ *  class MyComponent extends Component {
+ *      // Required arguments
+ *      @requiredAction doTheThing: (arg: number) => void;
+ *  }
+ * ```
+ */
+export function requiredAction(
+    classPrototype: any,
+    propertyName: string,
+): any {
+    const className = classPrototype.constructor.name;
+
+    // eslint-disable-next-line no-param-reassign
+    classPrototype[propertyName] = () => {
+        assert(`Missing required action in ${className}: ${propertyName}`);
+    };
+}
+
+/**
+ * Property decorator for optional action arguments in components.
+ * Provides a default action method that does nothing.
+ *
+ * Usage:
+ *
+ * ```ts
+ *  import { optionalAction } from 'ember-osf-web/utils/required-action';
+ *
+ *  class MyComponent extends Component {
+ *      // Optional arguments
+ *      @optionalAction doTheThing: (arg: number) => void;
+ *  }
+ * ```
+ */
+export function optionalAction(
+    classPrototype: any,
+    propertyName: string,
+): any {
+    // eslint-disable-next-line no-param-reassign
+    classPrototype[propertyName] = () => { /* empty */ };
+}
+
+/**
+ * Class decorator to set layout and styles in components.
+ *
+ * Same as ember-decorator's `@layout`, but with an additional optional `styles` param
+ */
+export function layout(template: unknown, styles?: unknown) {
+    return <T extends Newable<Component>>(ComponentSubclass: T) => {
+        /* eslint-disable no-param-reassign */
+        ComponentSubclass.prototype.layout = template;
+        ComponentSubclass.prototype.styles = styles;
+        /* eslint-enable no-param-reassign */
+        return ComponentSubclass;
+    };
+}
```

Comment:
This is nice. I worry a little about the name collision with ember-decorator's `@layout` causing possible confusion. That said, I don't have a better name. Maybe `@view`?

Reply:
I'm open to other names. `@view` is pretty clear and nicely evokes the MVC pattern (where C is for "component"), but I'm not sure about it given the confusing history of "views" in ember...

Reply:
The "V" in MVC is what I was thinking, but I agree it's too overloaded. What about `@ui`? (also probably too overloaded)

Reply:
I like that it looks like "oui"

Reply:
`@layout` works. Also, the fact that it just extends the interface of `ember-decorator`'s `@layout` with an optional parameter means accidental misuse is unlikely.